MarshalingPInvokeScanner doesn't see [DisableRuntimeMarshalling] attributes in any assembly except System.Runtime.dll

The MarshalingPInvokeScanner task doesn't see [DisableRuntimeMarshalling] attributes defined in any assembly except System.Runtime.dll (where the attribute is defined).

The problem is that this code:

https://github.com/dotnet/runtime/blob/035d2686771a79ee8b8f1b976f6ccc6dec427611/src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs#L116-L131

does not take into account that the attribute's constructor will be a MethodReference when referencing a constructor defined in another assembly.

Repro:

* Clone https://github.com/rolfbjarne/marshalling-scanner (including submodules)
* Run `make disable-runtime-marshalling-attribute`

This happens:

```
[...]
Assembly /Users/rolf/test/dotnet/marshalling-scanner/disable-runtime-marshalling-attribute/bin/Debug/net9.0/disable-runtime-marshalling-attribute.dll requires marshal-ilgen for method .Class1:DoSomething (first pass).
❌ Failed, because some assemblies are incompatible
make: *** [disable-runtime-marshalling-attribute] Error 1
```

--- a/src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs
+++ b/src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs
@@ -106,8 +106,8 @@ namespace MonoTargetsTasks
 
         private bool IsDisableRuntimeMarshallingAttribute (MetadataReader mdtReader, StringHandle ns, StringHandle name)
         {
-            return mdtReader.GetString(ns) == "System.Runtime.CompilerServices" &&
-                   mdtReader.GetString(name) == "DisableRuntimeMarshallingAttribute";
+            return mdtReader.StringComparer.Equals(ns, "System.Runtime.CompilerServices") &&
+                   mdtReader.StringComparer.Equals(name, "DisableRuntimeMarshallingAttribute");
         }
 
         private bool IsAssemblyIncompatible(string assyPath, MinimalMarshalingTypeCompatibilityProvider mmtcp)

--- a/src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs
+++ b/src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs
@@ -104,7 +104,7 @@ namespace MonoTargetsTasks
             }
         }
 
-        private bool IsDisableRuntimeMarshallingAttribute (MetadataReader mdtReader, StringHandle ns, StringHandle name)
+        private static bool IsDisableRuntimeMarshallingAttribute(MetadataReader mdtReader, StringHandle ns, StringHandle name)
         {
             return mdtReader.StringComparer.Equals(ns, "System.Runtime.CompilerServices") &&
                    mdtReader.StringComparer.Equals(name, "DisableRuntimeMarshallingAttribute");
@@ -133,7 +133,7 @@ namespace MonoTargetsTasks
                     if (IsDisableRuntimeMarshallingAttribute(mdtReader, td.Namespace, td.Name))
                         return false;
                 }
-                else if(attr.Constructor.Kind == HandleKind.MemberReference)
+                else if (attr.Constructor.Kind == HandleKind.MemberReference)
                 {
                     MemberReferenceHandle mrh = (MemberReferenceHandle)attr.Constructor;
                     MemberReference mr = mdtReader.GetMemberReference(mrh);

--- a/src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs
+++ b/src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs
@@ -104,6 +104,12 @@ namespace MonoTargetsTasks
             }
         }
 
+        private bool IsDisableRuntimeMarshallingAttribute (MetadataReader mdtReader, StringHandle ns, StringHandle name)
+        {
+            return mdtReader.GetString(ns) == "System.Runtime.CompilerServices" &&
+                   mdtReader.GetString(name) == "DisableRuntimeMarshallingAttribute";
+        }
+
         private bool IsAssemblyIncompatible(string assyPath, MinimalMarshalingTypeCompatibilityProvider mmtcp)
         {
             using FileStream file = new FileStream(assyPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
@@ -124,9 +130,19 @@ namespace MonoTargetsTasks
                     TypeDefinitionHandle tdh = md.GetDeclaringType();
                     TypeDefinition td = mdtReader.GetTypeDefinition(tdh);
 
-                    if (mdtReader.GetString(td.Namespace) == "System.Runtime.CompilerServices" &&
-                        mdtReader.GetString(td.Name) == "DisableRuntimeMarshallingAttribute")
+                    if (IsDisableRuntimeMarshallingAttribute(mdtReader, td.Namespace, td.Name))
                         return false;
+                }
+                else if(attr.Constructor.Kind == HandleKind.MemberReference)
+                {
+                    MemberReferenceHandle mrh = (MemberReferenceHandle)attr.Constructor;
+                    MemberReference mr = mdtReader.GetMemberReference(mrh);
+                    if (mr.Parent.Kind == HandleKind.TypeReference) {
+                        TypeReference tr = mdtReader.GetTypeReference((TypeReferenceHandle)mr.Parent);
+
+                        if (IsDisableRuntimeMarshallingAttribute(mdtReader, tr.Namespace, tr.Name))
+                            return false;
+                    }
                 }
             }
 

